draft/0005-impersonation.md

```diff
@@ -0,0 +1,333 @@
+3. `token_type`: REQUIRED. The type of the token. The value of this parameter must be `Bearer`.
+4. `expires_in`: REQUIRED. The lifetime in seconds of the access token.
+5. `scope`: OPTIONAL. The scopes of the access token.
+
+### Example token exchange
+
+The following example demonstrates a token exchange request and response. An authorized user is exchanging the subject token for a new access token that represents the identity of another user. Here's a non-normative example of the token exchange request:
+
+```
+POST /oidc/token HTTP/1.1
+Host: example.logto.app
+Content-Type: application/x-www-form-urlencoded
+
+grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Atoken-exchange
+&scope=openid%20offline_access%20profile
+&subject_token=accVkjcJyb4BWCxGsndESCJQbdFMogUC5PbRDqceLTC
+&subject_token_type=
+urn%3Aietf%3Aparams%3Aoauth%3Atoken-type%3Aaccess_token
+```
+
+Logto validates the `subject_token` and other parameters in the token exchange request. If the request is valid, Logto generates a new access token that represents the identity of the impersonated user and returns it in the response. Here's a non-normative example of the token exchange response:
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+Cache-Control: no-cache, no-store
+
+{
+  "access_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjMiLCJzY29wZSI6Im9wZW5pZCBvZmZsaW5lX2FjY2VzcyBwcm9maWxlIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJhY3QiOnsic3ViIjoiNDU2In19.H-X4W0mu-lJae27_rvsHAx5LpzsHTDYozkWDWzBrMoM",
+  "scope": "openid offline_access profile",
+  "issued_token_type": "urn:ietf:params:oauth:token-type:access_token",
+  "token_type": "Bearer",
+  "expires_in": 3600
+}
+```
+
+The payload of the access token:
+
+```json
+{
+  "sub": "123",
+  "scope": "openid offline_access profile",
+  "client_id": "client1",
+  "name": "John Doe",
+  "iat": 1516239022
+}
+```
+
+## Subject token
+
+The `subject_token` parameter in the token exchange request represents the identity of the party on behalf of whom the request is being made. The `subject_token` is a special one-time access token that is issued to the authorized user by Logto's management API. The `subject_token` is an opaque token that is not intended to be used directly by the client application. Instead, the client application exchanges the `subject_token` for a new access token that represents the identity of the impersonated user. Because the `subject_token` is only consumed by Logto as the authorization server, it does not need to be a JWT.
+
+### Issuing a subject token
+
+#### Request
+
+The client makes a request to the tenant's management API to issue a subject token. The following parameters are included in the HTTP request entity-body using the `application/json` format.
+
+1. `userId`: REQUIRED. The ID of the user on behalf of whom the request is being made.
```

Comment:
i'm thinking about if this is required. as we'll record the m2m app that requested the management api, it looks not required to specify a user id. i'm open to both since we can mark it optional later.

Reply:
This is required, otherwise, we are unable to identify the "subject" in token exchange.

Reply:
the subject can actually be a machine-to-machine app?

Reply:
No, we only support "user" here, it makes no sense to impersonate a machine to machine app, because it does not need UI to sign in.